src/cocotb/share/lib/vpi/VpiObj.cpp

```diff
@@ -63,48 +63,41 @@ int VpiArrayObjHdl::initialise(const std::string &name,
 
     /* After determining the range_idx, get the range and set the limits */
     vpiHandle iter = vpi_iterate(vpiRange, hdl);
-
-    s_vpi_value val;
-    val.format = vpiIntVal;
+    vpiHandle rangeHdl;
 
     if (iter != NULL) {
-        vpiHandle rangeHdl;
-        int idx = 0;
+        rangeHdl = vpi_scan(iter);
 
-        while ((rangeHdl = vpi_scan(iter)) != NULL) {
-            if (idx == range_idx) {
+#ifdef MODELSIM
+        for (int i = 0; i < range_idx; ++i) {
+            rangeHdl = vpi_scan(iter);
+            if (rangeHdl == NULL) {
                 break;
             }
-            ++idx;
         }
-
+#endif
+        vpi_free_object(iter);  // Need to free iterator since exited early
```

Comment:
If `rangeHdl` is null, the iterator handle is no longer valid so must not be freed here.